.github/workflows/build-and-test-differential.yaml

```diff
@@ -27,27 +27,41 @@ jobs:
     with:
       label: tag:require-cuda-build-and-test
 
-  build-and-test-differential:
+  prepare-build-and-test-differential:
+    runs-on: ubuntu-latest
     needs: [make-sure-label-is-present, make-sure-require-cuda-label-is-present]
-    if: ${{ needs.make-sure-label-is-present.outputs.result == 'true' }}
+    outputs:
+      cuda_build: ${{ steps.check-if-cuda-build-is-required.outputs.cuda_build }}
+    steps:
+      - name: Check if cuda-build is required
+        id: check-if-cuda-build-is-required
+        run: |
+          if ${{ needs.make-sure-require-cuda-label-is-present.outputs.result == 'true' }}; then
+            echo "cuda-build is required"
+            echo "cuda_build=true" >> $GITHUB_OUTPUT
+          else
+            echo "cuda-build is not required"
+            echo "cuda_build=false" >> $GITHUB_OUTPUT
+          fi
+        shell: bash
+      - name: Fail if build is not required
+        if: ${{ needs.make-sure-label-is-present.outputs.result != 'true' }}
+        run: exit 1
+
+  build-and-test-differential:
     runs-on: ${{ matrix.runner }}
     container: ${{ matrix.container }}${{ matrix.container-suffix }}
+    needs: prepare-build-and-test-differential
     strategy:
       fail-fast: false
       matrix:
         rosdistro:
           - humble
-        container-suffix:
-          - ""
-          - -cuda
         include:
           - rosdistro: humble
             container: ghcr.io/autowarefoundation/autoware:universe-devel
+            container-suffix: ""
             build-depends-repos: build_depends.repos
-          - container-suffix: -cuda
-            runner: codebuild-autoware-us-east-1-${{ github.run_id }}-${{ github.run_attempt }}-ubuntu-7.0-large
-            build-pre-command: taskset --cpu-list 0-5
-          - container-suffix: ""
             runner: ubuntu-latest
             build-pre-command: ""
```

Comment:
If we are splitting build-and-test-differential and build-and-test-differential-cuda as a job, then we probably don't have to use matrix

Reply:
I think the separation build-and-test-differential action made its maintenance more complex and the action logs harder to read.

Reply:
@mitsudome-r 
Matrix has been removed. Thank you!

Reply:
@xmfcx 

Thank you for your review. I see your point. Could you provide more details on how splitting `build-and-test-differential` makes maintenance more complex or action logs harder to read? I'll consider solutions to these issues!

Alternatively, if you have other ideas for addressing the problems described in the motivation, I'd like to hear them!

Reply:
Readeability got worse because GitHub dev team is lazy.

**See:** [Better log separation for composite actions #21276](https://github.com/orgs/community/discussions/21276)
And please vote and comment under the post 🙇

The substeps from composite actions are harder to read.

About the maintenance costs, you already have 2 versions of `clang-tidy-differential` for cuda and non-cuda. So what was extracted through separation of the action did not reduce the total lines of code in total by too much.

But added additional complexity due to introduced variables for the composite action.

And probably, handling the new cuda related tag could be done more easily without complicating the logic of the non-cuda.

## This PR

I like what you did with this PR. This was using a lot of codebuild time just to pull images on aws machines. (up to 3 minutes of compute time for each PR and each push) With this PR, if no tag, no codebuild instances will start. So thank you for implementing this.

Reply:
Thank you for the detailed explanation!

Regarding readability, it seems we can only hope for improvements from the GitHub dev team in the future.

As for maintenance, the attached materials were very informative for me. Thank you!!!